### PR TITLE
Validate that token is listed for reserve in KyberHintHandler

### PR DESCRIPTION
**Pull request by ayobuenavista**
Other improvements: MaskOut trade type with empty reserveId list is valid

### CODE REVIEW

File: contracts/sol6/KyberMatchingEngine.sol [hunk 1]
```diff
@@ -204,6 +209,17 @@ contract KyberMatchingEngine is KyberHintHandler, IKyberMatchingEngine, Withdraw
         (reserveAddress, , , ,) = kyberStorage.getReserveDetailsById(reserveId);
     }
 
+    function areAllReservesListed(
+        bytes32[] memory reserveIds,
+        IERC20 src,
+        IERC20 dest
+    ) internal override view returns (bool) {
```
Comment (ilanDoron) [inline contracts/sol6/KyberMatchingEngine.sol:216]: please use boolean values

returns(bool result)
and don't declare another one.

File: contracts/sol6/KyberMatchingEngine.sol [hunk 2]
```diff
@@ -94,14 +94,19 @@ contract KyberMatchingEngine is KyberHintHandler, IKyberMatchingEngine, Withdraw
...
             }
             if (dest == ETH_TOKEN_ADDRESS) {
                 (unpackedHint, ) = unpackT2THint(hint);
-                (tradeType, reserveIds, splitValuesBps, error) = parseHint(unpackedHint);
+                (tradeType, reserveIds, splitValuesBps, error) = parseHint(src, unpackedHint, true);
             }
         } else {
-            (tradeType, reserveIds, splitValuesBps, error) = parseHint(hint);
+            if (src == ETH_TOKEN_ADDRESS) {
+                (tradeType, reserveIds, splitValuesBps, error) = parseHint(dest, hint, false);
```
Comment (ilanDoron) [inline contracts/sol6/KyberMatchingEngine.sol:105]: we decided network should do this validation.
And it shouldn't trust matching engine

seems to me now the validation will be done twice for trade and get rate
right?

File: contracts/sol6/KyberMatchingEngine.sol (hunk 1 above)
Comment (ayobuenavista) [reply contracts/sol6/KyberMatchingEngine.sol]: Fixed

File: contracts/sol6/KyberMatchingEngine.sol (hunk 2 above)
Comment (ayobuenavista) [reply contracts/sol6/KyberMatchingEngine.sol]: Changed to external functions only.

File: contracts/sol6/KyberHintHandler.sol [hunk 3]
```diff
@@ -259,7 +274,7 @@ abstract contract KyberHintHandler is IKyberHint, Utils5 {
         hint = abi.encode(t2eHint, e2tHint);
     }
 
-    /// @notice Parses or decodes the token to ether or ether to token bytes hint
+    /// @notice Parses or decodes the Token to ETH or ETH to Token bytes hint
```
Comment (Anyhowclick) [inline contracts/sol6/KyberHintHandler.sol:277]: err, why change this ah?

Comment (Anyhowclick) [reply contracts/sol6/KyberHintHandler.sol:277]: I'll approve first, either Anton fix, or we fix in cleanup

**Pull request merged**